File: attention/scatter_pa_kv_cache_with_k_scale/op_host/op_api/scatter_pa_kv_cache_with_k_scale.cpp
/**
 * Copyright (c) 2026 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

#include "scatter_pa_kv_cache_with_k_scale.h"
#include "opdev/op_log.h"
#include "opdev/op_dfx.h"
#include "opdev/shape_utils.h"
#include "opdev/make_op_executor.h"
#include "opdev/platform.h"

using namespace op;

namespace l0op {
OP_TYPE_REGISTER(ScatterPaKvCacheWithKScale);

std::tuple<const aclTensor *, const aclTensor *, const aclTensor *> ScatterPaKvCacheWithKScale(const aclTensor *key,
    const aclTensor *value, aclTensor *keyCacheRef, aclTensor *valueCacheRef, const aclTensor *slotMapping,
    aclTensor *keyScale, aclTensor *keyScaleCacheRef, char *cacheLayoutOptional, aclOpExecutor *executor)
{
    L0_DFX(ScatterPaKvCacheWithKScale, key, value, keyCacheRef, valueCacheRef, slotMapping, keyScale, keyScaleCacheRef,
           cacheLayoutOptional);

    auto ret = ADD_TO_LAUNCHER_LIST_AICORE(
        ScatterPaKvCacheWithKScale,
        OP_INPUT(key, value, keyCacheRef, valueCacheRef, slotMapping, keyScale, keyScaleCacheRef),
        OP_OUTPUT(keyCacheRef, valueCacheRef, keyScaleCacheRef), OP_ATTR(cacheLayoutOptional));
    if (ret != ACLNN_SUCCESS) {
        OP_LOGE(ACLNN_ERR_PARAM_INVALID, "ADD_TO_LAUNCHER_LIST_AICORE failed.");
        return {nullptr, nullptr, nullptr};
    }
    return std::tuple<const aclTensor *, const aclTensor *, const aclTensor *>(keyCacheRef, valueCacheRef,
                                                                               keyScaleCacheRef);
}

} // namespace l0op

File: torch_extension/cann_ops_transformer/ops/scatter_pa_kv_cache_with_k_scale.py
# -----------------------------------------------------------------------------------------------------------
# Copyright (c) 2026 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# -----------------------------------------------------------------------------------------------------------

from typing import Optional, Tuple

import torch
import torch_npu
from torch.library import impl
from cann_ops_transformer.op_builder.builder import OpBuilder
from cann_ops_transformer.op_builder.builder import AS_LIBRARY


class ScatterPaKvCacheWithKScaleOpBuilder(OpBuilder):
    """
    torch_npu适配：scatter_pa_kv_cache_with_k_scale算子
    
    功能：将FP8格式的key/value和对应的key_scale更新到KV cache中
    
    输入：
        key: [num_tokens, num_head, k_head_size] FP8格式
        value: [num_tokens, num_head, v_head_size] FP8格式
        key_cache: [num_blocks, num_head, block_size, k_head_size] FP8格式
        value_cache: [num_blocks, num_head, block_size, v_head_size] FP8格式
        slot_mapping: [num_tokens] int32/int64
        key_scale: [num_tokens, num_head] float32
        key_scale_cache: [num_blocks, num_head, block_size, 1] float32
    
    输出（inplace更新）：
        key_cache, value_cache, key_scale_cache
    """
    
    def __init__(self):
        super(ScatterPaKvCacheWithKScaleOpBuilder, self).__init__("scatter_pa_kv_cache_with_k_scale")

    def sources(self):
        """C++源代码路径"""
        return ['ops/csrc/scatter_pa_kv_cache_with_k_scale.cpp']

    def schema(self) -> str:
        """PyTorch算子签名"""
        return (
            "scatter_pa_kv_cache_with_k_scale("
            "Tensor key, Tensor value, Tensor(a!) key_cache, Tensor(b!) value_cache, "
            "Tensor slot_mapping, Tensor key_scale, Tensor(c!) key_scale_cache, *, "
            "str cache_layout='BNBD') -> (Tensor(a!), Tensor(b!), Tensor(c!))"
        )

    def register_meta(self):
        """
        Meta实现：Shape/Dtype推导
        
        输出shape与输入cache相同（inplace更新）
        输出dtype与输入dtype相同
        """
        @impl(AS_LIBRARY, self.name, "Meta")
        def scatter_pa_kv_cache_with_k_scale_meta(
            key, value, key_cache, value_cache,
            slot_mapping, key_scale, key_scale_cache,
            *, cache_layout='BNBD'
        ):
            # 输出shape与输入cache相同
            key_cache_out = torch.empty(
                key_cache.size(),
                dtype=key_cache.dtype,
                device='meta'
            )
            
            value_cache_out = torch.empty(
                value_cache.size(),
                dtype=value_cache.dtype,
                device='meta'
            )
            
            key_scale_cache_out = torch.empty(
                key_scale_cache.size(),
                dtype=key_scale_cache.dtype,
                device='meta'
            )
            
            return (key_cache_out, value_cache_out, key_scale_cache_out)


# 实例化OpBuilder
scatter_pa_kv_cache_with_k_scale_op_builder = ScatterPaKvCacheWithKScaleOpBuilder()
op_module = scatter_pa_kv_cache_with_k_scale_op_builder.load()


@impl(AS_LIBRARY, scatter_pa_kv_cache_with_k_scale_op_builder.name, "PrivateUse1")
def scatter_pa_kv_cache_with_k_scale(
    key: torch.Tensor,
    value: torch.Tensor,
    key_cache: torch.Tensor,
    value_cache: torch.Tensor,
    slot_mapping: torch.Tensor,
    key_scale: torch.Tensor,
    key_scale_cache: torch.Tensor,
    *,
    cache_layout: Optional[str] = 'BNBD',
) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
    """
    torch_npu实现：scatter_pa_kv_cache_with_k_scale
    
    Args:
        key: [num_tokens, num_head, k_head_size] FP8格式，需要更新的key
        value: [num_tokens, num_head, v_head_size] FP8格式，需要更新的value
        key_cache: [num_blocks, num_head, block_size, k_head_size] FP8格式，被更新的cache
        value_cache: [num_blocks, num_head, block_size, v_head_size] FP8格式，被更新的cache
        slot_mapping: [num_tokens] int32/int64，每个token在cache中的偏移
        key_scale: [num_tokens, num_head] float32，需要更新的scale
        key_scale_cache: [num_blocks, num_head, block_size, 1] float32，被更新的scale cache
        cache_layout: str，cache布局格式（默认'BNBD'）
    
    Returns:
        tuple: (key_cache_out, value_cache_out, key_scale_cache_out)
        
    注意：
        1. 算子支持FP8_E5M2和FP8_E4M3FN两种dtype
        2. slot_mapping值必须在[0, num_blocks*block_size-1]范围内
        3. 超出范围的slot会被跳过（不更新对应cache）
    """
    return op_module.scatter_pa_kv_cache_with_k_scale(
        key, value, key_cache, value_cache,
        slot_mapping, key_scale, key_scale_cache,
        cache_layout
    )
